userd/settings.go

```diff
@@ -0,0 +1,222 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package userd
+
+import (
+	"fmt"
+	"os/exec"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/godbus/dbus"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/osutil"
+)
+
+// Timeout when the confirmation dialog for an xdg-setging
+// automatically closes. Keep in sync with the core snaps
+// xdg-settings wrapper which also sets this value to 300.
+var defaultConfirmDialogTimeout = 300 * time.Second
+
+const settingsIntrospectionXML = `
+<interface name="org.freedesktop.DBus.Peer">
+	<method name='Ping'>
+	</method>
+	<method name='GetMachineId'>
+               <arg type='s' name='machine_uuid' direction='out'/>
+	</method>
+</interface>
+<interface name='io.snapcraft.Settings'>
+	<method name='Check'>
+		<arg type='s' name='setting' direction='in'/>
+		<arg type='s' name='check' direction='in'/>
+		<arg type='s' name='result' direction='out'/>
+	</method>
+	<method name='Get'>
+		<arg type='s' name='setting' direction='in'/>
+		<arg type='s' name='result' direction='out'/>
+	</method>
+	<method name='Set'>
+		<arg type='s' name='setting' direction='in'/>
+		<arg type='s' name='value' direction='in'/>
+	</method>
+</interface>`
+
+// FIXME: allow setting default-url-scheme-handler ?
+var settingsWhitelist = []string{
+	"default-web-browser",
+}
+
+// FIXME: too restrictive?
+var allowedSettingsValues = regexp.MustCompile(`^[a-zA-Z0-9._]+$`)
```

Comment:
This should allow `<snap command>.desktop` I think (and no `_`). validate.go has validAppName and https://github.com/snapcore/snapd/pull/4399 has some code to add a new ValidAppName() that you could use. Eg, check if ends with `.desktop`, the use the helpers after splitting on '_' (and '.').